rgw: DefaultRetention requires either Days or Years

liuchang0812 opened the pull request:
Either Days or Years must be specified, but not both.

see https://docs.aws.amazon.com/AmazonS3/latest/API/Type_API_DefaultRetention.html

Signed-off-by: Chang Liu <liuchang0812@gmail.com>

@cbodley @yehudasa mind taking a look?

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

tchaikov commented on src/rgw/rgw_object_lock.h:
```diff
@@ -106,7 +106,11 @@ class RGWObjectLock
+    // You can't specify both at the same time.
+    // see https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTObjectLockConfiguration.html
+    return (get_years() > 0 || get_days() > 0) && 
+	   !(get_years() > 0 && get_days() > 0);
```
could be
```c++
return get_years() > 0 != get_days() > 0;
```

tchaikov commented on src/rgw/rgw_object_lock.h:
```diff
@@ -106,7 +106,10 @@ class RGWObjectLock
+    // DefaultRetention requires either Days or Years.
+    // You can't specify both at the same time.
+    // see https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTObjectLockConfiguration.html
+    return (get_years() > 0) != (get_days() > 0);
```
nit, no need to put parentheses around the statement after `return`.

liuchang0812 replied:
thanks, there is a warning if we remove those parentheses. 

``` 
/home/liuchang/WorkSpace/ceph/src/rgw/rgw_object_lock.h:112:24: warning: suggest parentheses around comparison in operand of ‘!=’ [-Wparentheses]
     return get_years() > 0 != get_days() > 0;
            ~~~~~~~~~~~~^~~
```

ivancich commented:
rgw/(non-)verify: http://pulpito.ceph.com/ivancich-2019-09-11_01:23:37-rgw-wip-eric-testing-1-distro-basic-smithi/
rgw/verify: http://pulpito.ceph.com/ivancich-2019-09-13_17:22:14-rgw-wip-eric-testing-1-bisect-6-distro-basic-smithi/

ivancich commented:
@liuchang0812 Can you create a tracker for this and cross-link them? It passed my QA and I'm ready to merge, but so we need the tracker so the backports can be tracked. Thanks!

liuchang0812 commented:
> @liuchang0812 Can you create a tracker for this and cross-link them? It passed my QA and I'm ready to merge, but so we need the tracker so the backports can be tracked. Thanks!

done, https://tracker.ceph.com/issues/41838, and I have updated pr's commit message and rebased to master.

liuchang0812 commented:
jenkins test make check

cbodley commented:
jenkins test make check
